aten/src/ATen/core/dispatch/OperatorEntry.cpp

```diff
@@ -153,51 +156,91 @@ const KernelFunction& OperatorEntry::computeDispatchTableEntry(const c10::Dispat
   return computeDispatchTableEntryWithDebug(dispatcher, dispatch_key).first.kernel;
 }
 
+bool OperatorEntry::hasKernelForDispatchKeySet(DispatchKeySet ks) const {
+  for (auto k : ks) {
+    if (kernels_.find(k) != kernels_.end()) {
+      return true;
+    }
+  }
+  return false;
+}
+
+c10::optional<const AnnotatedKernel*> OperatorEntry::getKernelForDispatchKey(DispatchKey dispatch_key) const{
+  auto kern_it = kernels_.find(dispatch_key);
+  if (kern_it != kernels_.end()) {
+    TORCH_INTERNAL_ASSERT(!kernels_.at(dispatch_key).empty());
+    TORCH_INTERNAL_ASSERT(kernels_.at(dispatch_key).front().kernel.isValid());
+    return c10::make_optional(&kernels_.at(dispatch_key).front());
+  }
+  return c10::nullopt;
+}
+
 std::pair<const AnnotatedKernel&, const char*> OperatorEntry::computeDispatchTableEntryWithDebug(const c10::Dispatcher& dispatcher, DispatchKey dispatch_key) const {
   auto dispatch_ix = static_cast<uint8_t>(dispatch_key);
   // [Note] DispatchTable computation
   // dispatchTable contains entries for runtime dispatch keys.
   // For any dispatch key, it'll pick a kernel using the following order:
   //  (1) Use kernel if it's directly registered to this key
   //  (2) Handle runtime keys that have kernels available from alias keys
-  //    (2.1) Use kernel from DispatchKey::Autograd if available
-  //    (2.2) For autograd backend keys, we use kernel from alias Math key (catchAll will be moved to Math)
+  //    (2.1) Use kernel from DispatchKey::Math if available.
+  //          For autograd keys, we only use kernel from Math when there's no direct registration
+  //          to its corresponding backend key.
+  //          For AutogradOther, we eagerly return ambiguousAutogradOtherKernel_ if there's registration to any of
+  //          its backends and ask backend extender to request a decicated Autograd key for the backend.
+  //          See Note [Ambiguity in AutogradOther kernel] for more details.
+  //    (2.2) Use kernel from DispatchKey::Autograd if available
+  //    (2.3) Special logic to handle catchAll for Autograd keys
+  //          For autograd backend keys, we use kernel from alias Math key (catchAll will be moved to Math)
```

Comment:
Also I think it's worth noting in the comments that the implementation relies on the invariant that  `computeDispatchTableEntryWithDebug()` will be called first for backend keys, then for autograd keys (or more precisely, for a given backend, it'll be called for that backend's autograd key *after* the backend key). In fact it might be worth specifically referencing the logic at the bottom of `updateDispatchTable_` that we're relying on to preserve the invariant.